MaxText/layers/linears.py

```diff
@@ -156,11 +152,36 @@ def compute_dot_general(inputs, kernel, axis, contract_ind):
 
     contract_ind = tuple(range(0, len(axis)))
     output = compute_dot_general(inputs, kernel, axis, contract_ind)
+    return output
+
+
+class DenseWithNormBias(DenseGeneral):
```

Comment:
If we are going to have a separate layer Let's make this layer very specific for DeepSeek Gating (e.g. reflecting in the name).

We can also not have a separate layer, and for DeepSeek use `DenseGeneral` for DeepSeek, then  add a bias (while keeping previous weights) and apply scaling.

Both options are good in my opinion.

Reply:
Yeah, sounds good. The tricky thing is this is just a dense matmul, not like gating, which also need to combines routing logic. I think we have 2 good options, 1) have 'DenseDeepseek', which will have some duplicate logics with 'DenseGeneral', as score_func change is in the middle (before bias). 2) (this could be a later version) refactor routing logics into a new 'MoeGate', this will need ckpt structure changed (probably just an extra layer), since MoeGate will call DenseGeneral or DenseDeepSeek and then select topk. I will do a refactor option 1 first, and follow up with 2nd option.

Reply:
Let's quickly chat in person later? and then I will do the refactor once we are aligned. Thank you!